### PR TITLE
Warn about asterisk being used in media type.

### LINKED ISSUE #181: plugin-types directive should not accept wildcards
Currently, we do accept wildcards: https://cspvalidator.org/#headerValue%5B%5D=plugin-types+*%2F*&strategy=intersection

### PR DESCRIPTION
**Pull request by shekyan**
fixes #181

### CODE REVIEW

File: src/main/java/com/shapesecurity/salvation/Constants.java [hunk 1]
```diff
@@ -14,7 +14,8 @@
             + "|scripts|top-navigation)$");
     public static final Pattern requireSriForEnumeratedTokenPattern = Pattern
             .compile("^(?:script|style)$", Pattern.CASE_INSENSITIVE);
-    public static final Pattern mediaTypePattern = Pattern.compile("^(?<type>[^/]+)/(?<subtype>[^/]+)$");
+    public static final Pattern mediaTypePattern = Pattern.compile("^(?<type>[a-zA-Z0-9!#$%^&*\\-_+{}|'.`~]+)/(?<subtype>[a-zA-Z0-9!#$%^&*\\-_+{}|'.`~]+)$");
+    // public static final Pattern mediaTypePattern = Pattern.compile("[a-zA-Z0-9!#$%^&\\*-_\\+{}\\|'.`~]+/[a-zA-Z0-9!#$%^&\\*-_\\+{}\\|'.`~]+");
```
Comment (kingthorin) [inline src/main/java/com/shapesecurity/salvation/Constants.java]: Drop this commented one?

File: src/main/java/com/shapesecurity/salvation/Parser.java [hunk 2]
```diff
@@ -252,6 +252,8 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
                     if (mediaTypes.isEmpty()) {
                         this.error(token, "The media-type-list must contain at least one media-type.");
                         throw INVALID_MEDIA_TYPE_LIST;
+                    } else if (mediaTypes.stream().anyMatch(x -> x.matchesTypeOrSubType("*"))) {
+                        this.warn(token,"`*` is allowed character in media type which will be literally matched. Make sure it not an attempt to match any.");
```
Comment (kingthorin) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: > Make sure it not an attempt to match any.

Missing an "is" in there. Also if the user/dev does want to define "any" what should they be using? Or is it not possible? Either way, some clarity or a suggestion might help...

Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/Parser.java]: Thanks! Addressed comments.

File: src/main/java/com/shapesecurity/salvation/directiveValues/MediaType.java [hunk 3]
```diff
@@ -25,6 +25,10 @@ public boolean matchesMediaType(@Nonnull MediaType mediaType) {
         return this.type.equalsIgnoreCase(mediaType.type) && this.subtype.equalsIgnoreCase(mediaType.subtype);
     }
 
+    public boolean matchesTypeOrSubType(@Nonnull String str) {
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/MediaType.java]: I don't like this function. It's not generally useful. Just use `x.type.equals("*") || x.subtype.equals("*")`.

File: src/main/java/com/shapesecurity/salvation/Constants.java [hunk 4]
```diff
@@ -14,7 +14,7 @@
             + "|scripts|top-navigation)$");
     public static final Pattern requireSriForEnumeratedTokenPattern = Pattern
             .compile("^(?:script|style)$", Pattern.CASE_INSENSITIVE);
-    public static final Pattern mediaTypePattern = Pattern.compile("^(?<type>[^/]+)/(?<subtype>[^/]+)$");
+    public static final Pattern mediaTypePattern = Pattern.compile("^(?<type>[a-zA-Z0-9!#$%^&*\\-_+{}|'.`~]+)/(?<subtype>[a-zA-Z0-9!#$%^&*\\-_+{}|'.`~]+)$");
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/Constants.java]: Add comment with link to grammar/spec and the name of the production this is supposed to match.

Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/Constants.java]: Addressed

**Pull request merged**